hbase-common/src/main/java/org/apache/hadoop/hbase/io/crypto/tls/X509Util.java

```diff
@@ -0,0 +1,310 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.io.crypto.tls;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.security.GeneralSecurityException;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.NoSuchAlgorithmException;
+import java.security.Security;
+import java.security.cert.PKIXBuilderParameters;
+import java.security.cert.X509CertSelector;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.CertPathTrustManagerParameters;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedTrustManager;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.exceptions.X509Exception;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility code for X509 handling Default cipher suites: Performance testing done by Facebook
+ * engineers shows that on Intel x86_64 machines, Java9 performs better with GCM and Java8 performs
+ * better with CBC, so these seem like reasonable defaults.
+ */
+@InterfaceAudience.Private
+public class X509Util {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509Util.class);
+
+  // Config
+  static final String CONFIG_PREFIX = "hbase.rpc.tls.";
+  static final String TLS_CONFIG_PROTOCOL = CONFIG_PREFIX + "protocol";
+  static final String TLS_CONFIG_KEYSTORE_LOCATION = CONFIG_PREFIX + "keystore.location";
+  static final String TLS_CONFIG_KEYSTORE_TYPE = CONFIG_PREFIX + "keystore.type";
+  static final String TLS_CONFIG_KEYSTORE_PASSWORD = CONFIG_PREFIX + "keystore.password";
+  static final String TLS_CONFIG_TRUSTSTORE_LOCATION = CONFIG_PREFIX + "truststore.location";
+  static final String TLS_CONFIG_TRUSTSTORE_TYPE = CONFIG_PREFIX + "truststore.type";
+  static final String TLS_CONFIG_TRUSTSTORE_PASSWORD = CONFIG_PREFIX + "truststore.password";
+  static final String TLS_CONFIG_CLR = CONFIG_PREFIX + "clr";
+  static final String TLS_CONFIG_OCSP = CONFIG_PREFIX + "ocsp";
+
+  public static final String HBASE_NETTY_RPCSERVER_TLS_ENABLED =
+    "hbase.netty.rpcserver.tls.enabled";
```

Comment:
it may at the very least make sense to have a separate client vs server config for enabling TLS.  i ran into this in my tests, I think it's impossible to gracefully transition a live cluster to TLS without a separate config. 

Even with supportplaintext enabled, the moment you enable tls on a regionserver it will fail to connect to HMasters because they aren't expecting SSL. But if you upgrade HMaster first, they will fail to communicate with regionservers. So we really need a way to separately enable rpcserver and rpcclient tls

Reply:
So what would be the right process of upgrading?
1. Upgrade Masters first and enable supportplaintext
2. RSs are able to connect with plaintext SASL
3. Upgrade RSs in a rolling restart fashion with supportplaintext
4. Restarted region servers connect via TLS
5. Clients can connect master/RS via plaintext
6. Upgrade clients to TLS
7. Disable supportplaintext on master / restart
8. Disable supportplaintext on RSs / restart

Does it make sens?
I haven't verified this process, just dumping my brain.

Reply:
So unfortunately that process will not work. Since there is a single config for enabling both client and server side TLS. The moment you do step 1, the HMasters will try sending RPCs to RSs using TLS. But the RSs will not have TLS enabled and will not understand the encrypted requests. Example is OpenRegion/CloseRegion requests. 

That said, I agree that would be a reasonable approach to aim for. I think we need to provide a separate config for clients to make that work though. In which case the final step would be to restart HMaster one more time with client enabled. 

If region replication is used, might need to do RSs in two phases (server enabled then client) too.

Reply:
I suggest providing two config: `hbase.client.netty.tls.enabled` and `hbase.server.netty.tls.enabled`

I think there's a case to be made for removing netty from the name. Netty is the only implementation in master branch I believe, so it might be confusing eventually.

Reply:
Thanks for the explanation @bbeaudreault , it makes perfect sense. I'll make the change.